File: backend/engine/IndicatorEngine.ts
// ============================================================
// Indicator Engine — Ported from frontend IndicatorEngine.ts
// Pure calculation, no browser dependencies
// ============================================================

import type { Candle, IndicatorSnapshot } from './types.js';
import { emptyIndicatorSnapshot } from './types.js';

export function ema(values: number[], period: number): (number | null)[] {
  if (period <= 0 || values.length < period) {
    return Array(values.length).fill(null);
  }

  const result: (number | null)[] = Array(values.length).fill(null);
  const multiplier = 2.0 / (period + 1);
  const seed = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
  result[period - 1] = seed;

  let previousEMA = seed;
  for (let i = period; i < values.length; i++) {
    const currentEMA = (values[i] - previousEMA) * multiplier + previousEMA;
    result[i] = currentEMA;
    previousEMA = currentEMA;
  }

  return result;
}

function rsiValue(avgGain: number, avgLoss: number): number {
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function rsi(values: number[], period: number = 14): (number | null)[] {
  if (values.length <= period) {
    return Array(values.length).fill(null);
  }

  const result: (number | null)[] = Array(values.length).fill(null);
  let gains = 0;
  let losses = 0;

  for (let i = 1; i <= period; i++) {
    const change = values[i] - values[i - 1];
    if (change >= 0) gains += change;
    else losses += Math.abs(change);
  }

  let avgGain = gains / period;
  let avgLoss = losses / period;
  result[period] = rsiValue(avgGain, avgLoss);

  if (values.length <= period + 1) return result;

  for (let i = period + 1; i < values.length; i++) {
    const change = values[i] - values[i - 1];
    const gain = Math.max(change, 0);
    const loss = Math.max(-change, 0);
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    result[i] = rsiValue(avgGain, avgLoss);
  }

  return result;
}

export function macd(values: number[]): { macd: (number | null)[]; signal: (number | null)[] } {
  const ema12 = ema(values, 12);
  const ema26 = ema(values, 26);
  const macdLine: (number | null)[] = Array(values.length).fill(null);

  for (let i = 0; i < values.length; i++) {
    if (ema12[i] !== null && ema26[i] !== null) {
      macdLine[i] = ema12[i]! - ema26[i]!;
    }
  }

  const compactMACD = macdLine.filter((v): v is number => v !== null);
  const compactSignal = ema(compactMACD, 9);
  const signalLine: (number | null)[] = Array(values.length).fill(null);
  let compactIdx = 0;

  for (let i = 0; i < macdLine.length; i++) {
    if (macdLine[i] !== null) {
      if (compactIdx < compactSignal.length) {
        signalLine[i] = compactSignal[compactIdx];
      }
      compactIdx++;
    }
  }

  return { macd: macdLine, signal: signalLine };
}

function previousValue(values: (number | null)[]): number | null {
  const compact = values.filter((v): v is number => v !== null);
  return compact.length >= 2 ? compact[compact.length - 2] : null;
}

export function snapshot(candles: Candle[]): IndicatorSnapshot {
  if (candles.length === 0) return { ...emptyIndicatorSnapshot };

  const closes = candles.map(c => c.close);
  const volumes = candles.map(c => c.volume);
  const latest = candles[candles.length - 1];

  const ema9Values = ema(closes, 9);
  const ema21Values = ema(closes, 21);
  const ema50Values = ema(closes, 50);
  const rsiValues = rsi(closes);
  const macdValues = macd(closes);

  const avgVol20: number | null = volumes.length >= 20
    ? volumes.slice(-20).reduce((a, b) => a + b, 0) / 20
    : null;

  const recent = candles.slice(-20);
  const support = Math.min(...recent.map(c => c.low));
  const resistance = Math.max(...recent.map(c => c.high));

  return {
    ema9: ema9Values[ema9Values.length - 1] ?? null,
    ema21: ema21Values[ema21Values.length - 1] ?? null,
    ema50: ema50Values[ema50Values.length - 1] ?? null,
    rsi14: rsiValues[rsiValues.length - 1] ?? null,
    previousRSI14: previousValue(rsiValues),
    macd: macdValues.macd[macdValues.macd.length - 1] ?? null,
    macdSignal: macdValues.signal[macdValues.signal.length - 1] ?? null,
    previousMACD: previousValue(macdValues.macd),
    previousMACDSignal: previousValue(macdValues.signal),
    averageVolume20: avgVol20,
    currentVolume: latest.volume,
    support,
    resistance,
  };
}

File: ios/AI Crypto Analyzer/IndicatorEngine.swift
import Foundation

enum IndicatorEngine {
    static func ema(values: [Double], period: Int) -> [Double?] {
        guard period > 0, values.count >= period else {
            return Array(repeating: nil, count: values.count)
        }

        var result = Array<Double?>(repeating: nil, count: values.count)
        let multiplier = 2.0 / Double(period + 1)
        let seed = values.prefix(period).reduce(0, +) / Double(period)
        result[period - 1] = seed

        var previousEMA = seed
        for index in period..<values.count {
            let currentEMA = (values[index] - previousEMA) * multiplier + previousEMA
            result[index] = currentEMA
            previousEMA = currentEMA
        }

        return result
    }

    static func rsi(values: [Double], period: Int = 14) -> [Double?] {
        guard values.count > period else {
            return Array(repeating: nil, count: values.count)
        }

        var result = Array<Double?>(repeating: nil, count: values.count)
        var gains = 0.0
        var losses = 0.0

        for index in 1...period {
            let change = values[index] - values[index - 1]
            if change >= 0 {
                gains += change
            } else {
                losses += abs(change)
            }
        }

        var averageGain = gains / Double(period)
        var averageLoss = losses / Double(period)
        result[period] = rsiValue(averageGain: averageGain, averageLoss: averageLoss)

        guard values.count > period + 1 else { return result }

        for index in (period + 1)..<values.count {
            let change = values[index] - values[index - 1]
            let gain = max(change, 0)
            let loss = max(-change, 0)
            averageGain = ((averageGain * Double(period - 1)) + gain) / Double(period)
            averageLoss = ((averageLoss * Double(period - 1)) + loss) / Double(period)
            result[index] = rsiValue(averageGain: averageGain, averageLoss: averageLoss)
        }

        return result
    }

    static func macd(values: [Double]) -> (macd: [Double?], signal: [Double?]) {
        let ema12 = ema(values: values, period: 12)
        let ema26 = ema(values: values, period: 26)
        var macdLine = Array<Double?>(repeating: nil, count: values.count)

        for index in values.indices {
            guard let fast = ema12[index], let slow = ema26[index] else { continue }
            macdLine[index] = fast - slow
        }

        let compactMACD = macdLine.compactMap { $0 }
        let compactSignal = ema(values: compactMACD, period: 9)
        var signalLine = Array<Double?>(repeating: nil, count: values.count)
        var compactIndex = 0

        for index in macdLine.indices where macdLine[index] != nil {
            if compactIndex < compactSignal.count {
                signalLine[index] = compactSignal[compactIndex]
            }
            compactIndex += 1
        }

        return (macdLine, signalLine)
    }

    static func snapshot(for candles: [Candle]) -> IndicatorSnapshot {
        let closes = candles.map(\.close)
        let volumes = candles.map(\.volume)
        guard let latest = candles.last else { return IndicatorSnapshot() }

        let ema9Values = ema(values: closes, period: 9)
        let ema21Values = ema(values: closes, period: 21)
        let ema50Values = ema(values: closes, period: 50)
        let rsiValues = rsi(values: closes)
        let macdValues = macd(values: closes)

        let averageVolume20: Double?
        if volumes.count >= 20 {
            averageVolume20 = volumes.suffix(20).reduce(0, +) / 20
        } else {
            averageVolume20 = nil
        }

        let recent = candles.suffix(20)
        let support = recent.map(\.low).min()
        let resistance = recent.map(\.high).max()

        return IndicatorSnapshot(
            ema9: ema9Values.last ?? nil,
            ema21: ema21Values.last ?? nil,
            ema50: ema50Values.last ?? nil,
            rsi14: rsiValues.last ?? nil,
            previousRSI14: previousValue(in: rsiValues),
            macd: macdValues.macd.last ?? nil,
            macdSignal: macdValues.signal.last ?? nil,
            previousMACD: previousValue(in: macdValues.macd),
            previousMACDSignal: previousValue(in: macdValues.signal),
            averageVolume20: averageVolume20,
            currentVolume: latest.volume,
            support: support,
            resistance: resistance
        )
    }

    private static func rsiValue(averageGain: Double, averageLoss: Double) -> Double {
        if averageLoss == 0 { return 100 }
        let relativeStrength = averageGain / averageLoss
        return 100 - (100 / (1 + relativeStrength))
    }

    private static func previousValue(in values: [Double?]) -> Double? {
        let compact = values.compactMap { $0 }
        guard compact.count >= 2 else { return nil }
        return compact[compact.count - 2]
    }
}

